### PR TITLE
Integrating compiler Gym into Nevergrad

### PR DESCRIPTION
**Pull request by teytaud**
## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Motivation and Context / Related issue

Variant of #1132 

## Checklist

- [x] The documentation is up-to-date with the changes I made.
- [x] I have read the [**CONTRIBUTING**](https://facebookresearch.github.io/nevergrad/contributing.html) document and completed the CLA (see [**CLA**](https://facebookresearch.github.io/nevergrad/contributing.html#contributor-license-agreement-cla)).
- [x] All tests passed, and additional code has been covered with new tests.

### CODE REVIEW

File: nevergrad/benchmark/plotting.py [hunk 1]
```diff
@@ -313,6 +313,10 @@ def create_plots(
             print("\n# new case #", fixed, case)
             casedf = df.select(**dict(zip(fixed, case)))
             data_df = FightPlotter.winrates_from_selection(casedf, fight_descriptors, num_rows=num_rows)
+            # We also want a bigger dataframe with more columns for writing a text overview.
+            data_df_big = FightPlotter.winrates_from_selection(
+                casedf, fight_descriptors, num_rows=2, num_cols=100
+            )
```
Comment (jrapin) [inline nevergrad/benchmark/plotting.py:319]: this should not be called twice because it's slow, can we get the big one, and then cut the smaller one out of it?

File: nevergrad/benchmark/plotting.py [hunk 2]
```diff
@@ -435,7 +449,11 @@ def __init__(
         self._ax.grid(True, which="both")
         self._overlays: tp.List[tp.Any] = []
         legend_infos: tp.List[LegendInfo] = []
-        for optim_name in sorted_optimizers:
+        for optim_name in (
+            sorted_optimizers[:1] + sorted_optimizers[-12:]
```
Comment (jrapin) [inline nevergrad/benchmark/plotting.py:453]: the first one and the last 12? I dont get this

File: nevergrad/benchmark/plotting.py [hunk 3]
```diff
@@ -552,6 +571,28 @@ def make_data(df: pd.DataFrame, normalized_loss: bool = False) -> tp.Dict[str, t
...
+        optim_vals: dict
+            dict of losses obtained by a given optimizer.
+        """
+        best_performance: tp.Dict[int, tp.Any] = defaultdict(lambda: (float("inf"), "none"))
+        for optim in optim_vals.keys():
+            for i, l in zip(optim_vals[optim]["budget"], optim_vals[optim]["loss"]):
+                if l < best_performance[i][0]:
+                    best_performance[i] = (l, optim)
+
+        with open(output_filepath, "w") as f:
```
Comment (jrapin) [inline nevergrad/benchmark/plotting.py:589]:
```suggestion
        with output_filepath.open("w") as f:
```

File: nevergrad/benchmark/plotting.py [hunk 4]
```diff
@@ -560,9 +601,14 @@ def save(self, output_filepath: tp.PathLike) -> None:
...
             path where the figure must be saved
         """
-        self._fig.savefig(
-            str(output_filepath), bbox_extra_artists=self._overlays, bbox_inches="tight", dpi=_DPI
-        )
+        try:  # Let us catch errors due to too many DPIs.
+            self._fig.savefig(
+                str(output_filepath), bbox_extra_artists=self._overlays, bbox_inches="tight", dpi=_DPI
+            )
+        except ValueError:
```
Comment (jrapin) [inline nevergrad/benchmark/plotting.py:608]: what happened?

File: nevergrad/benchmark/plotting.py [hunk 5]
```diff
@@ -621,7 +667,11 @@ def __init__(self, winrates_df: pd.DataFrame) -> None:
...
     @staticmethod
     def winrates_from_selection(
-        df: utils.Selector, categories: tp.List[str], num_rows: int = 5
+        df: utils.Selector,
+        categories: tp.List[str],
+        num_rows: int = 5,
+        num_cols: int = 30,
+        restricted_to_complete: bool = False,
     ) -> pd.DataFrame:
         """Creates a fight plot win rate data out of the given run dataframe,
```
Comment (jrapin) [inline nevergrad/benchmark/plotting.py:676]: this will need to be changed to the other PR

File: nevergrad/benchmark/test_experiments.py [hunk 6]
```diff
@@ -32,6 +32,10 @@ def test_experiments_registry(name: str, maker: tp.Callable[[], tp.Iterator[expe
     if "_pgan" in name and os.environ.get("CIRCLECI", False):
         raise SkipTest("Too slow in CircleCI")
 
+    # CompilerGym sends http requests that CircleCI does not like.
+    if "compiler" in name and os.environ.get("CIRCLECI", False):
+        raise SkipTest("No HTTP request in CircleCI")
```
Comment (jrapin) [inline nevergrad/benchmark/test_experiments.py:37]: please put this in the code of the function that does not work instead (if CIRCLECI then raise SkipTest). Relying on the name is never a good idea

File: nevergrad/benchmark/test_experiments.py [hunk 7]
```diff
@@ -40,20 +44,24 @@ def test_experiments_registry(name: str, maker: tp.Callable[[], tp.Iterator[expe
     if all(x in name for x in ["image", "quality"]) and platform.system() == "Windows":
         raise SkipTest("Image quality not guaranteed on Windows.")
 
+    # ANM does not work under Windows.
+
     # Basic test.
+    print(f"Testing {name}")
```
Comment (jrapin) [inline nevergrad/benchmark/test_experiments.py:50]:
```suggestion
```

File: nevergrad/functions/gym/test_gym_multi.py [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+        assert env_name in gym_multi.GymMulti.env_names, f"{env_name} unknown!"
+        assert env_name not in gym_multi.NO_LENGTH, f"{env_name} in no length and in ng_gym!"
+    for env_name in gym_multi.GUARANTEED_GYM_ENV_NAMES:
+        assert env_name in gym_multi.GymMulti.env_names, f"{env_name} should be guaranteed!"
+    assert len(gym_multi.GYM_ENV_NAMES) >= 26 or os.name == "nt"
+
+
+def test_compiler_gym() -> None:
+    if os.environ.get("CIRCLECI", False):
+        raise SkipTest("No HTTP request in CircleCI")
```
Comment (jrapin) [inline nevergrad/functions/gym/test_gym_multi.py:24]: as mentioned above, this should go in the function directly, then no need to copy it everywhere

File: nevergrad/functions/gym/test_gym_multi.py [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+    results = [func(np.zeros(func.dimension)) for _ in range(40)]
+    assert min(results) != max(results), "Roulette should not be deterministic."
+    candidate = func.parametrization.sample()
+    results = [func.evaluation_function(candidate) for _ in range(40)]
+    assert min(results) != max(results), "Roulette should not be deterministic."
+
+
+@pytest.mark.parametrize("name", gym_multi.GYM_ENV_NAMES)
+def test_run_gym_multi(name) -> None:
+    if os.name != "nt" and all(np.random.randint(2, size=3, dtype=bool)):
```
Comment (jrapin) [inline nevergrad/functions/gym/test_gym_multi.py:43]: same, if this does not work on nt, then raise UnsupportedExperiment in the function

File: nevergrad/functions/gym/test_gym_multi.py [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+        func = gym_multi.GymMulti(
+            name,
+            control,
+            neural_factor=(None if (control == "linear" or "conformant" in control) else 1),
+            randomized=bool(np.random.randint(2)),
+        )
+        y = func.parametrization.sample()
+        func(y.value)
+        if "stac" in control:  # Let's check if the memory works.
+            func(y.value)
```
Comment (jrapin) [inline nevergrad/functions/gym/test_gym_multi.py:60]: how is that testing that the memory works? shouldnt you compare the values?

File: nevergrad/optimization/optimizerlib.py [hunk 11]
```diff
@@ -2519,3 +2519,11 @@ def __init__(
...
     ) -> None:
         super().__init__(_MSR, locals())
+
+
+ProgD13 = ConfSplitOptimizer(
+    num_optims=13, progressive=True, multivariate_optimizer=NoisyDiscreteOnePlusOne
+).set_name("ProgD13", register=True)
+Prog13 = ConfSplitOptimizer(
+    num_optims=13, progressive=True, multivariate_optimizer=OptimisticNoisyOnePlusOne
+).set_name("Prog13", register=True)
```
Comment (jrapin) [inline nevergrad/optimization/optimizerlib.py:2529]: can it go to experimental variants?

File: nevergrad/functions/gym/gym_multi.py [hunk 12]
```diff
@@ -0,0 +1,924 @@
...
+import os
+import typing as tp
+import gym
+
+compiler_gym_present = True
+try:
+    import compiler_gym  # pylint: disable=unused-import
+    from compiler_gym.envs import CompilerEnv
+except ImportError:
+    compiler_gym_present = False
```
Comment (jrapin) [inline nevergrad/functions/gym/gym_multi.py:17]: move the import to the function, raise ng.errors.UnsupportedExperiment if not working

File: nevergrad/functions/gym/gym_multi.py [hunk 13]
```diff
@@ -0,0 +1,924 @@
...
+
+import nevergrad as ng
+
+if os.name != "nt":
+    import gym_anm  # pylint: disable=unused-import
```
Comment (jrapin) [inline nevergrad/functions/gym/gym_multi.py:22]: same

File: nevergrad/functions/gym/gym_multi.py [hunk 14]
```diff
@@ -0,0 +1,924 @@
...
+        a3 = np.asarray(env.action_space.sample())
+        a1 = a1 + a2 + a3
+        if hasattr(a1, "size"):
+            try:
+                assert a1.size < 15000  # type: ignore
+            except:
+                assert a1.size() < 15000  # type: ignore
+        GYM_ENV_NAMES.append(e.id)
+    except:
+        pass
```
Comment (jrapin) [inline nevergrad/functions/gym/gym_multi.py:47]: please no inline code, move this to a function

**Review by jrapin — CHANGES_REQUESTED** [review]

File: nevergrad/benchmark/plotting.py (hunk 2 above)
Comment (teytaud) [reply nevergrad/benchmark/plotting.py:453]: This includes the 12 best, and the worst method.

This is for having an eye on bias in noisy optimization. 

Consider a case in which the expected loss value is always >= 0 (equal to 0 at the optimum).
In the noisy case, you will still get a sudden decrease of the 12 best methods at the last x-axis value: because we select methods with the best losses at the last x-axis value.

With also the worst, you get something more interpretable: you see that this is just noise.

File: nevergrad/functions/gym/gym_multi.py (hunk 14 above)
Comment (teytaud) [reply nevergrad/functions/gym/gym_multi.py]: done.

File: nevergrad/benchmark/plotting.py (hunk 4 above)
Comment (teytaud) [reply nevergrad/benchmark/plotting.py:608]: Sometimes plots are too big.
I add a print.

File: nevergrad/optimization/optimizerlib.py (hunk 11 above)
Comment (teytaud) [reply nevergrad/optimization/optimizerlib.py:2538]: Entirely removed.

File: nevergrad/benchmark/plotting.py (hunk 3 above)
Comment (teytaud) [reply nevergrad/benchmark/plotting.py]: Does not work because << nevergrad/benchmark/plotting.py:589: error: Item "str" of "Union[str, Path]" has no attribute "open"
 >>. What is the goal of this proposed change ?

File: nevergrad/functions/gym/gym_multi.py (hunk 12 above)
Comment (teytaud) [reply nevergrad/functions/gym/gym_multi.py]: done.

File: nevergrad/functions/gym/gym_multi.py (hunk 13 above)
Comment (teytaud) [reply nevergrad/functions/gym/gym_multi.py]: done.

File: nevergrad/functions/gym/test_gym_multi.py (hunk 10 above)
Comment (teytaud) [reply nevergrad/functions/gym/test_gym_multi.py:60]: We just check that nothing crashes. No specific value expected here.

File: nevergrad/benchmark/plotting.py (hunk 1 above)
Comment (teytaud) [reply nevergrad/benchmark/plotting.py]: Ok: only one now. 
We print only the 30 best then, which is ok for now.

File: nevergrad/benchmark/experiments.py [hunk 15]
```diff
@@ -1133,6 +1135,254 @@ def rocket(seed: tp.Optional[int] = None, seq: bool = False) -> tp.Iterator[Expe
 
 
 @registry.register
+def ng_full_gym(
```
[the remark this answers is not in the document]
Comment (teytaud) [reply nevergrad/benchmark/experiments.py]: Done.

File: nevergrad/benchmark/experiments.py [hunk 16]
```diff
@@ -1138,7 +1140,6 @@ def rocket(seed: tp.Optional[int] = None, seq: bool = False) -> tp.Iterator[Expe
                             yield xp
 
 
-@registry.register
```
Comment (jrapin) [inline nevergrad/benchmark/experiments.py:1141]: why is this removed? Can we remove the xp?

File: nevergrad/benchmark/experiments_gym.py [hunk 17]
```diff
@@ -0,0 +1,295 @@
...
+    if conformant:
+        controls = ["stochastic_conformant"]
+    for control in controls:
+        neural_factors = [None] if (conformant or control == "linear") else ([1] if "memory" in control else [3 if big else 1])  # type: ignore
+        for neural_factor in neural_factors:
+            for name in env_names:
+                try:
+                    func = GymMulti(name, control=control, neural_factor=neural_factor, randomized=randomized)
+                except MemoryError:
+                    pass
```
Comment (jrapin) [inline nevergrad/benchmark/experiments_gym.py:85]: This is dangerous because then it falls back to the previous function, you want `continue` instead

Comment (teytaud) [reply nevergrad/benchmark/experiments_gym.py:85]: Agree yes.

File: nevergrad/benchmark/experiments.py (hunk 16 above)
Comment (teytaud) [reply nevergrad/benchmark/experiments.py:1141]: I unregistered so that it is no more in the dashboard because mujoco = reproducibility issues, but we might keep it for now as there are users who like it. Yeah I guess unregistering was wrong.

Comment (jrapin) [reply nevergrad/benchmark/experiments.py:1141]: users can use it if it's unregistered :s so it's better to keep it registered or delete altogether

**Comment by jrapin** [comment]
There are changes which have been updated in #1134, please merge #1134 into master first then master in this branch

**Comment by teytaud** [comment]
> There are changes which have been updated in #1134, please merge #1134 into master first then master in this branch

Done.

File: nevergrad/benchmark/experiments.py (hunk 16 above)
Comment (teytaud) [reply nevergrad/benchmark/experiments.py]: Thanks for having brought them back.

**Review by jrapin — APPROVED** [review]

File: nevergrad/functions/gym/test_multigym.py [hunk 18]
```diff
@@ -0,0 +1,61 @@
...
+    func = multigym.GymMulti(
+        name,
+        control,
+        neural_factor=(None if (control == "linear" or "conformant" in control) else 1),
+        randomized=bool(np.random.randint(2)),
+    )
+    y = func.parametrization.sample()
+    func(y.value)
+    if "stac" in control:  # Let's check if the memory works.
+        func(y.value)
```
Comment (jrapin) [inline nevergrad/functions/gym/test_multigym.py:61]: @teytaud shouldnt this be tested against the previous value ? This looks like it tests nothing right now

Comment (teytaud) [reply nevergrad/functions/gym/test_multigym.py]: I add a test on the output value.
